### DISCUSSION TITLE
Raw sql migrations

### DISCUSSION THREAD

**Discussion by a0nguyen** [emmett-framework/emmett#348] [Q&A] [answered]
In some case, I prefer using Raw sql to perform migrations, for ex:
- complex schema(s),
- taking advantage of vendor specific extensions to SQL (both DDL and DML)
- richer / additional data types
- using vendor specific procedural code (i.e. PL/pgSQL anonymous blocks in PostgreSQL)
etc

Is there a way to do raw sql migrations in emmett?

**Comment by gi0baro** [comment, marked as answer]
@a0nguyen you can generate an empty migration [as described in the docs](https://emmett.sh/docs/2.2.x/orm/migrations#empty-migrations-and-operations) using the relative command

    emmett migrations new -m "Some name here"

then you can use the `executesql` command to run whatever SQL query you prefer in the migration `up` and `down` methods, eg:

```python
def up(self):
    # runs when migration is applied
    self.db.executesql("ALTER TABLE ...")

def down(self):
    # runs when downgraded to previous
    self.db.executesql("ALTER TABLE...")
```

you can also avoid to write `down` commands, but ensure to raise some error so it's clear migration cannot be downgraded.
